### PR TITLE
On Windows, set name of Window Class, closes #2977

### LINKED ISSUE #2977: On Windows, allow to customize the name of the Window Class (current value hard coded is "Window Class")
Mirror of issue opened on Tauri forks of winit : https://github.com/tauri-apps/tao/issues/769

**Is your feature request related to a problem? Please describe.**
Currently, on Windows, all windows created by winit have the same class Window Class, it would be neet to allow for customization of the Windows' window class name. One use case: some windows custom hotkey managers leverage window class names to select windows...

**Describe the solution you'd like**
- In `src\platform\windows.rs` add a `with_window_classname`
- In `src\platform_impl\windows\mod.rs`, in `PlatformSpecificWindowBuilderAttributes` add a `window_classname` attribute and in `impl Default for PlatformSpecificWindowBuilderAttributes` add a default value for `window_classname` of `Window Class` (non breaking change)

**Describe alternatives you've considered**
No alternative, unless simply dumping those tools and find one using other characteristics of windows...

**Would you want to assign yourself to implement this feature?**
- [X] Yes
- [ ] No

### PR DESCRIPTION
**Pull request by geraudloup**
allow to customize it instead of current value hard coded "Window Class"

- [X] Tested on all platforms changed
- [X] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
- [X] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [X] Created or updated an example program if it would help users understand this functionality
- [X] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

### CODE REVIEW

**Comment by madsmtm** [comment]
Linking your related issue: https://github.com/rust-windowing/winit/pull/2978

File: CHANGELOG.md [hunk 1]
```diff
@@ -8,6 +8,7 @@ And please only add new entries to the top of this list, right below the `# Unre
 
 # Unreleased
 
+- On Windows, added `WindowBuilderExtWindows::with_window_classname` and changed the `register_window_class` definition and call in `init` to use newly defined Windows Platform Attribute `window_classname`.
```
Comment (msiglreith) [inline CHANGELOG.md:11]:
```suggestion
- On Windows, added `WindowBuilderExtWindows::with_class_name` to customize the internal class name.
```

File: src/platform/windows.rs [hunk 2]
```diff
@@ -223,6 +223,9 @@ pub trait WindowBuilderExtWindows {
     /// Whether show or hide the window icon in the taskbar.
     fn with_skip_taskbar(self, skip: bool) -> WindowBuilder;
 
+    /// Customize the window class name.
+    fn with_window_classname<S: Into<String>>(self, classname: S) -> WindowBuilder;
```
Comment (msiglreith) [inline src/platform/windows.rs:227]: let's just call it `with_class_name` and adjust the attribute name as well.

File: src/platform_impl/windows/window.rs [hunk 3]
```diff
@@ -1102,7 +1102,7 @@ where
 {
     let title = util::encode_wide(&attributes.title);
 
-    let class_name = register_window_class::<T>();
+    let class_name = register_window_class::<T>(&pl_attribs.window_classname);
```
Comment (msiglreith) [inline src/platform_impl/windows/window.rs:1105]:
```suggestion
    let class_name = util::encode_wide(&pl_attribs.class_name);
    register_window_class::<T>(&class_name);
```

File: src/platform_impl/windows/window.rs [hunk 4]
```diff
@@ -1189,8 +1189,8 @@ where
     Ok(initdata.window.unwrap())
 }
 
-unsafe fn register_window_class<T: 'static>() -> Vec<u16> {
-    let class_name = util::encode_wide("Window Class");
+unsafe fn register_window_class<T: 'static>(window_classname: &String) -> Vec<u16> {
```
Comment (msiglreith) [inline src/platform_impl/windows/window.rs:1192]: With the change suggested above we can simplify it a bit to
```suggestion
unsafe fn register_window_class<T: 'static>(class_name: &[u16]) {
```

**Review by msiglreith — CHANGES_REQUESTED** [review]
thanks for the pull request!
I would request a few smaller changes, otherwise looks fine to me

File: FEATURES.md [hunk 5]
```diff
@@ -119,6 +119,7 @@ If your PR makes notable changes to Winit's features, please update this section
 
 ## Platform
 ### Windows
+* Setting the name of the Window Class used When creating the window
```
Comment (msiglreith) [inline FEATURES.md:122]:
```suggestion
* Setting the name of the internal window class
```

**Review by msiglreith — APPROVED** [review]

**Comment by daxpedda** [comment]
Some additional documentation on what a "Window Class" is and what it's used for would be nice.

**Pull request merged**